File: python/agentsecrets/testing/mock.py
"""Mock client for testing.

``MockAgentSecrets`` is a drop-in replacement for ``AgentSecrets`` that
records every ``call()`` and ``spawn()`` invocation without touching
the proxy or keychain.  It lets downstream developers write fast,
isolated tests for code that depends on the SDK.

Usage::

    from agentsecrets.testing import MockAgentSecrets

    mock = MockAgentSecrets()
    response = mock.call("https://api.example.com", bearer="KEY")

    assert len(mock.calls) == 1
    assert mock.calls[0]["url"] == "https://api.example.com"
    assert mock.calls[0]["bearer"] == "KEY"
    # No credential values are ever stored.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from ..models import AgentSecretsResponse, SpawnResult


@dataclass
class CallRecord:
    """What was passed to a ``call()`` invocation.

    Records injection **key names** — never values.
    """

    url: str
    method: str = "GET"
    bearer: str | None = None
    basic: str | None = None
    header: dict[str, str] | None = None
    query: dict[str, str] | None = None
    body_field: dict[str, str] | None = None
    form_field: dict[str, str] | None = None
    body: Any = None


@dataclass
class SpawnRecord:
    """What was passed to a ``spawn()`` invocation."""

    command: list[str] = field(default_factory=list)
    capture: bool = True


class MockAgentSecrets:
    """Drop-in replacement for ``AgentSecrets`` in tests.

    Parameters
    ----------
    default_response:
        The ``AgentSecretsResponse`` returned by ``call()`` / ``async_call()``
        unless overridden per-call.
    default_spawn_result:
        The ``SpawnResult`` returned by ``spawn()`` / ``spawn_async()``.
    """

    def __init__(
        self,
        *,
        default_response: AgentSecretsResponse | None = None,
        default_spawn_result: SpawnResult | None = None,
    ) -> None:
        self.calls: list[CallRecord] = []
        self.spawns: list[SpawnRecord] = []

        self._default_response = default_response or AgentSecretsResponse(
            status_code=200,
            headers={"Content-Type": "application/json"},
            body=b'{"ok": true}',
        )
        self._default_spawn = default_spawn_result or SpawnResult(exit_code=0)

    # ------------------------------------------------------------------
    # call / async_call
    # ------------------------------------------------------------------

    def call(
        self,
        url: str,
        *,
        method: str = "GET",
        body: Any = None,
        headers: dict[str, str] | None = None,
        bearer: str | None = None,
        basic: str | None = None,
        header: dict[str, str] | None = None,
        query: dict[str, str] | None = None,
        body_field: dict[str, str] | None = None,
        form_field: dict[str, str] | None = None,
        agent_id: str | None = None,
        timeout: float = 30.0,
    ) -> AgentSecretsResponse:
        """Record the call and return the default mock response."""
        self.calls.append(CallRecord(
            url=url,
            method=method,
            bearer=bearer,
            basic=basic,
            header=header,
            query=query,
            body_field=body_field,
            form_field=form_field,
            body=body,
        ))
        return self._default_response

    async def async_call(
        self,
        url: str,
        **kwargs: Any,
    ) -> AgentSecretsResponse:
        """Async variant — same recording, same response."""
        return self.call(url, **kwargs)

    # ------------------------------------------------------------------
    # spawn / spawn_async
    # ------------------------------------------------------------------

    def spawn(
        self,
        command: list[str],
        *,
        capture: bool = True,
        timeout: float | None = None,
    ) -> SpawnResult:
        """Record the spawn and return the default mock result."""
        self.spawns.append(SpawnRecord(command=command, capture=capture))
        return self._default_spawn

    async def spawn_async(
        self,
        command: list[str],
        **kwargs: Any,
    ) -> SpawnResult:
        """Async variant — same recording, same result."""
        return self.spawn(command, **kwargs)

    # ------------------------------------------------------------------
    # Context manager (no-op)
    # ------------------------------------------------------------------

    def __enter__(self) -> MockAgentSecrets:
        return self

    def __exit__(self, *exc: object) -> None:
        pass

    def close(self) -> None:
        """No-op — nothing to clean up."""

File: javascript/src/testing/mock.ts
/**
 * AgentSecrets SDK — Mock Client for Testing
 *
 * Drop-in replacement for AgentSecrets in tests.
 * Records every call() and spawn() without touching the proxy or keychain.
 * Mirrors MockAgentSecrets from the Python SDK (testing/mock.py).
 *
 * @example
 * import { MockAgentSecrets } from "@the-17/agentsecrets-sdk/testing";
 *
 * const mock = new MockAgentSecrets();
 * const response = await mock.call({ url: "https://api.stripe.com/v1/balance", bearer: "STRIPE_KEY" });
 *
 * assert(mock.calls.length === 1);
 * assert(mock.calls[0].url === "https://api.stripe.com/v1/balance");
 * assert(mock.calls[0].bearer === "STRIPE_KEY");
 * // No credential values are ever stored.
 */

import { AgentSecretsResponse, type CallOptions, type SpawnOptions, type SpawnResult } from "../types.js";

export interface CallRecord {
  url: string;
  method: string;
  bearer?: string;
  basic?: string;
  header?: Record<string, string>;
  query?: Record<string, string>;
  bodyField?: Record<string, string>;
  formField?: Record<string, string>;
  body?: unknown;
  agentId?: string;
}

export interface SpawnRecord {
  command: string[];
  capture: boolean;
}

export class MockAgentSecrets {
  readonly calls: CallRecord[] = [];
  readonly spawns: SpawnRecord[] = [];
  /** Workspace names passed to withWorkspace() — for assertions in tests */
  readonly _workspaceSwitches: string[] = [];
  /** Project names passed to withProject() — for assertions in tests */
  readonly _projectSwitches: string[] = [];

  private readonly defaultResponse: AgentSecretsResponse;
  private readonly defaultSpawnResult: SpawnResult;

  constructor(opts: {
    defaultResponse?: AgentSecretsResponse;
    defaultSpawnResult?: SpawnResult;
  } = {}) {
    this.defaultResponse = opts.defaultResponse ?? new AgentSecretsResponse({
      statusCode: 200,
      headers: { "content-type": "application/json" },
      body: new TextEncoder().encode('{"ok": true}'),
    });
    this.defaultSpawnResult = opts.defaultSpawnResult ?? {
      exitCode: 0,
      stdout: "",
      stderr: "",
    };
  }

  async call<T = unknown>(opts: CallOptions): Promise<AgentSecretsResponse<T>> {
    // Development-time guard: warn if body looks like it contains a real credential
    if (process.env["NODE_ENV"] !== "production" && opts.body !== undefined) {
      const bodyStr = typeof opts.body === "string"
        ? opts.body
        : JSON.stringify(opts.body);
      const credPattern = /sk_live_|sk-proj-|ghp_/;
      if (credPattern.test(bodyStr)) {
        console.warn(
          "[AgentSecrets MockAgentSecrets] Warning: body may contain a real credential value. " +
          "Pass secret KEY NAMES, not values. e.g. bearer: \"STRIPE_KEY\" not \"sk_live_...\"."
        );
      }
    }
    const record: CallRecord = { url: opts.url, method: opts.method ?? "GET" };
    if (opts.bearer    !== undefined) record.bearer    = opts.bearer;
    if (opts.basic     !== undefined) record.basic     = opts.basic;
    if (opts.header    !== undefined) record.header    = opts.header;
    if (opts.query     !== undefined) record.query     = opts.query;
    if (opts.bodyField !== undefined) record.bodyField = opts.bodyField;
    if (opts.formField !== undefined) record.formField = opts.formField;
    if (opts.body      !== undefined) record.body      = opts.body;
    if (opts.agentId   !== undefined) record.agentId   = opts.agentId;
    this.calls.push(record);
    return this.defaultResponse as AgentSecretsResponse<T>;
  }

  async spawn(opts: SpawnOptions): Promise<SpawnResult> {
    this.spawns.push({ command: opts.command, capture: opts.capture ?? true });
    return this.defaultSpawnResult;
  }

  async isProxyRunning(): Promise<boolean> { return true; }

  /**
   * Alias for call() — matches AgentSecrets.asyncCall() signature.
   */
  asyncCall<T = unknown>(opts: CallOptions): Promise<AgentSecretsResponse<T>> {
    return this.call<T>(opts);
  }

  /**
   * Mock withWorkspace — runs fn() immediately without switching anything.
   * Records the workspace name so tests can assert it was used.
   */
  async withWorkspace<T>(name: string, fn: () => Promise<T>): Promise<T> {
    this._workspaceSwitches.push(name);
    return fn();
  }

  /**
   * Mock withProject — runs fn() immediately without switching anything.
   * Records the project name so tests can assert it was used.
   */
  async withProject<T>(name: string, fn: () => Promise<T>): Promise<T> {
    this._projectSwitches.push(name);
    return fn();
  }

  close(): void { /* no-op */ }

  async [Symbol.asyncDispose](): Promise<void> { this.close(); }
}
